core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -0,0 +1,363 @@
+        if (map.isEmpty()) {
+            return;
+        }
+
+        synchronized (ctx) {
+            final Map<String, String> oldMap = getMap(ctx);
+            final Map<String, String> newMap;
+            if (oldMap.isEmpty()) {
+                newMap = new Object2ObjectOpenHashMap<>(map);
+            } else {
+                newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
+                newMap.putAll(oldMap);
+                newMap.putAll(map);
+            }
+            ctx.setAttr(MAP, newMap);
+        }
+    }
+
+    /**
+     * Copies the specified thread-local {@link MDC} property to the specified {@link RequestContext}.
+     *
+     * @param ctx the {@link RequestContext}
+     * @param key the key of the thread-local {@link MDC} property to copy
+     */
+    public static void copy(RequestContext ctx, String key) {
+        checkState(delegate != null, ERROR_MESSAGE);
+        put(ctx, key, delegate.get(key));
+    }
+
+    /**
+     * Copies all thread-local {@link MDC} properties to the specified {@link RequestContext}.
+     *
+     * @param ctx the {@link RequestContext}
+     */
+    public static void copyAll(RequestContext ctx) {
+        checkState(delegate != null, ERROR_MESSAGE);
+        putAll(ctx, firstNonNull(delegate.getCopyOfContextMap(), Collections.emptyMap()));
+    }
+
+    /**
+     * Unbinds the specified request-scoped {@link MDC} property from the specified {@link RequestContext}.
+     *
+     * @param ctx the {@link RequestContext}
+     * @param key the key of the request-scoped {@link MDC} property to unbind
+     */
+    public static void remove(RequestContext ctx, String key) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(key, "key");
+
+        synchronized (ctx) {
+            final Map<String, String> oldMap = getMap(ctx);
+            if (!oldMap.containsKey(key)) {
+                return;
+            }
+
+            final Map<String, String> newMap;
+            if (oldMap.size() == 1) {
+                newMap = Collections.emptyMap();
+            } else {
+                newMap = new Object2ObjectOpenHashMap<>(oldMap);
+                newMap.remove(key);
+            }
+            ctx.setAttr(MAP, newMap);
+        }
+    }
+
+    /**
+     * Unbinds all request-scoped {@link MDC} properties from the specified {@link RequestContext}.
+     *
+     * @param ctx the {@link RequestContext}
+     */
+    public static void clear(RequestContext ctx) {
+        requireNonNull(ctx, "ctx");
+
+        synchronized (ctx) {
+            final Map<String, String> oldMap = getMap(ctx);
+            if (!oldMap.isEmpty()) {
+                ctx.setAttr(MAP, Collections.emptyMap());
+            }
+        }
+    }
+
+    private static Map<String, String> getMap(RequestContext ctx) {
+        final Map<String, String> map;
+        if (ctx instanceof ClientRequestContext) {
+            map = ((ClientRequestContext) ctx).ownAttr(MAP);
```

Comment:
I'm wondering we should add `owdAttr` to `RequestContext` so that we do not have to check the instance all the time. 🤔

Reply:
Yeah, and `computeOwnAttrIfAbsent()`, etc. :smile: